sqlglot/optimizer/annotate_types.py

```diff
@@ -104,6 +108,78 @@ class TypeAnnotator:
         },
     }
 
+    FUNCTION_TYPES = {
```

Comment:
What about functions like `COERCE`, in which the return type depends on the input type?

Reply:
it should be an optional lambda

Reply:
Ah, we're expecting "custom annotator functions" for this.

Reply:
why do you need FUNCTION_TYPES dict at all? why not just use annotators?

Reply:
That's what I'm working on, pushing in a bit.